### PR TITLE
pipeline synchronous pipeline definition interface

### PR DESCRIPTION
**Pull request by kakj-go**
#### What type of this PR
/kind feature

#### What this PR does / why we need it:
Unified management of pipeline definitions, and other business logic can be implemented through definitions in the future

#### Which issue(s) this PR fixes:
erda-issue: [erda-issue](https://erda.cloud/erda/dop/projects/387/issues/all?id=206835&issueFilter__urlQuery=eyJpdGVyYXRpb25JRHMiOls1MDZdLCJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIiwiUkVTT0xWRUQiXSwiYXNzaWduZWVJRHMiOlsiMTAwMDU2MCJdfQ%3D%3D&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=506&type=TASK)

### CODE REVIEW

File: apistructs/pipeline_definition.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apistructs
+
+import (
+	"time"
+)
+
+type PipelineDefinitionProcessRequest struct {
```
Comment (sfwn) [inline apistructs/pipeline_definition.go:21]: move to definition types.

File: apistructs/pipeline_definition.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+type PipelineDefinitionProcessResponseData struct {
+	ID              uint64         `json:"id"`
+	PipelineSource  PipelineSource `json:"pipelineSource"`
+	PipelineYmlName string         `json:"pipelineYmlName"`
+	PipelineYml     string         `json:"pipelineYml"`
+	VersionLock     uint64         `json:"versionLock"`
+}
+
+type PipelineDefinitionProcessResponse struct {
```
Comment (sfwn) [inline apistructs/pipeline_definition.go:39]: no need

File: apistructs/pipeline_definition.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	PipelineSource        PipelineSource           `json:"pipelineSource"`
+	PipelineYmlName       string                   `json:"pipelineYmlName"`
+	PipelineYml           string                   `json:"pipelineYml"`
+	SnippetConfig         *SnippetConfig           `json:"snippetConfig"`
+	VersionLock           uint64                   `json:"versionLock"`
+	IsDelete              bool                     `json:"isDelete"`
+	PipelineCreateRequest *PipelineCreateRequestV2 `json:"pipelineCreateRequest"`
+}
+
+type PipelineDefinitionProcessResponseData struct {
```
Comment (sfwn) [inline apistructs/pipeline_definition.go:31]: no need

File: apistructs/pipeline_definition.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	SnippetConfig         *SnippetConfigOrder      `json:"snippetConfig"`
+	VersionLock           uint64                   `json:"versionLock"`
+	PipelineCreateRequest *PipelineCreateRequestV2 `json:"pipelineCreateRequest"`
+
+	IsDelete    bool       `json:"isDelete"`
+	TimeCreated *time.Time `json:"timeCreated,omitempty"`
+	TimeUpdated *time.Time `json:"timeUpdated,omitempty"`
+}
+
+type GetPipelineDefinitionVersionRequest struct {
```
Comment (sfwn) [inline apistructs/pipeline_definition.go:58]: no need, no complex struct, use pb is enough.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: apistructs/pipeline_yml.go [hunk 5]
```diff
@@ -123,15 +123,21 @@ type SnippetLabel struct {
 	Value string `json:"value"`
 }
 
+type SnippetConfigOrder struct {
```
Comment (sfwn) [inline apistructs/pipeline_yml.go:126]: add ut for order.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/definition/get_version_lock.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+import (
+	"fmt"
+
+	"github.com/erda-project/erda-proto-go/core/pipeline/definition/pb"
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/modules/pipeline/services/apierrors"
+)
+
+type GetPipelineDefinitionVersionRequestBo struct {
```
Comment (sfwn) [inline modules/pipeline/providers/definition/get_version_lock.go:25]: what is `Bo`

File: modules/pipeline/providers/definition/define.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package definition
+
+import (
+	"github.com/erda-project/erda/modules/pipeline/providers/definition/db"
+)
+
+type PipelineDefinitionManage struct {
```
Comment (sfwn) [inline modules/pipeline/providers/definition/define.go:21]: No need. Use provider directly.

File: modules/pipeline/providers/definition/provider.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+// +provider
+type provider struct {
+	Cfg      *config
+	Log      logs.Logger
+	Register transport.Register  `autowired:"service-register"`
+	MySQL    mysqlxorm.Interface `autowired:"mysql-xorm"`
+
+	definitionService *definitionService
+	svc               *PipelineDefinitionManage
```
Comment (sfwn) [inline modules/pipeline/providers/definition/provider.go:37]: `definitionService` is `svc`. No need `PipelineDefinitionManage`.

File: modules/pipeline/spec/pipeline_definition.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+package spec
+
+import (
+	"time"
+
+	"github.com/erda-project/erda/apistructs"
+)
+
+type PipelineDefinition struct {
```
Comment (sfwn) [inline modules/pipeline/spec/pipeline_definition.go:23]: move db spec to corresponding db pkg's define.go.

File: apistructs/pipeline_yml.go (hunk 5 above)
Comment (kakj-go) [reply apistructs/pipeline_yml.go:126]: Already, the toString method of SnippetConfig strcut will cover

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/spec/pipeline_definition_relation.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spec
+
+import (
+	"time"
+)
+
+type PipelineDefinitionRelation struct {
```
Comment (sfwn) [inline modules/pipeline/spec/pipeline_definition_relation.go:21]: remove it.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/definition/get_version_lock.go (hunk 6 above)
Comment (kakj-go) [reply modules/pipeline/providers/definition/get_version_lock.go:25]: image

File: .erda/migrations/pipeline/20210903-create-pipeline-definitions.sql [hunk 11]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE `pipeline_definitions` (
+  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT COMMENT '自增id',
+  `pipeline_yml` mediumtext NOT NULL COMMENT 'yml 的定义',
+  `extra` mediumtext NOT NULL COMMENT '定义详情',
```
Comment (sfwn) [inline .erda/migrations/pipeline/20210903-create-pipeline-definitions.sql:4]: adjust the order.
`id`, `source`, `name`, `snippet`, `yml`, `extra`

File: modules/pipeline/providers/definition/get_version_lock.go (hunk 6 above)
Comment (sfwn) [inline modules/pipeline/providers/definition/get_version_lock.go]: remove "Bo", use pkg to distinguish.

File: modules/pipeline/providers/definition/get_version_lock.go [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+	PipelineSource  apistructs.PipelineSource
+	PipelineYmlName string
+}
+
+type GetPipelineDefinitionVersionResponseBo struct {
```
Comment (sfwn) [inline modules/pipeline/providers/definition/get_version_lock.go]: Response doesn't need transfer.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/definition/get_version_lock.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+package definition
+
+import (
+	"github.com/erda-project/erda-proto-go/core/pipeline/definition/pb"
+	"github.com/erda-project/erda/modules/pipeline/providers/definition/transform_type"
+	"github.com/erda-project/erda/modules/pipeline/services/apierrors"
+)
+
+func (qm *definitionService) GetPipelineDefinitionVersionLock(bo *transform_type.GetPipelineDefinitionVersion) (*pb.PipelineDefinitionProcessVersionResponse, error) {
```
Comment (sfwn) [inline modules/pipeline/providers/definition/get_version_lock.go:23]: receiver name use `s` uniformly.

File: modules/pipeline/providers/definition/process.go [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+			defer wait.Done()
+			handlerError = append(handlerError, handler(*pipelineDefinition, *pipelineYml))
+		}()
+	}
+	wait.Wait()
+
+	return handlerError
+}
+
+func (qm *definitionService) ProcessPipelineDefinition(ctx context.Context, bo *transform_type.PipelineDefinitionProcess) (*transform_type.PipelineDefinitionProcessResult, error) {
```
Comment (sfwn) [inline modules/pipeline/providers/definition/process.go:75]: use `s`

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — APPROVED** [review]

**Pull request merged**